### ISSUE TITLE
Sparse ndarray class

### ISSUE THREAD

**Issue by mrocklin** [scipy/scipy#8162]
While sitting with @stefanv last week we discussed (and then @stefanv started implementing) sparse array classes that use the same code as the existing matrix classes, but instead satisfy the ndarray interface (only up to two dimensions).  For example slicing out rows would result in 1d arrays rather than 2d matrices, the `*` operator means elementwise multiply rather than matrix multiply, etc..

This would be valuable for a number of communities and, in particular, anyone trying to use sparse arrays and numpy arrays interchangably, as happens in Pandas, Scikit-Learn, and Dask.array.

I thought I'd raise this issue to track this work (and also to subtly prod @stefanv).

Also cc @NelleV and @njsmith

**Comment by stefanv** [comment]
Yes, I was speaking about this with @jarrodmillman, @njsmith, and @nellev, and they said that @rgommers had some thoughts on the topic, so roping him in too.

I am, indeed, interested in doing the work on this.  I could use help with ideas for a test suite of ndarray functionality *limited to 2D and lower*.

**Comment by perimosocordiae** [comment]
Feel free to use part or all of the prototype I wrote a while back: https://github.com/perimosocordiae/sparray

I implemented an arbitrary-dimension sparse array type similar to the existing COO format by storing an array of linearized indices, the nonzero data, and the shape. I tried to follow the ndarray API as much as possible.

Also take a look at @ev-br's attempt (https://github.com/ev-br/sparr) which uses C++ maps similar to the existing DOK format.

**Comment by perimosocordiae** [comment]
A few thoughts:
 * Supporting arbitrary dimensions isn't much harder than supporting <= 2. I would caution against a design that bakes 2D shapes into the code too deeply.
 * Try to avoid ambiguity about if/when duplicate entries are allowed, and in general when internal sparse storage invariants need to be maintained. Same goes for stored zeros. These issues have caused a lot of headaches in the existing spmatrix codebase.
 * It'll remain important to give users a choice of the sparse storage format for efficiency reasons, but we'll want a design that doesn't guarantee that all operations on a given format will produce results of the same format. This is especially relevant when considering operations that produce results with different dimensions than the input, like slicing or extracting a diagonal.

**Comment by mrocklin** [comment]
I believe that the immediate goal was to just reuse the same data structures to get 1d and 2d arrays with minimal effort.

**Comment by stefanv** [comment]
Yes, I think the idea, for now, is to do the minimum necessary to get a more familiar API to work with across different projects.  Once that is in place, we should be able to refactor at our leasure, and eventually expand to higher ND.  It's probably not a lot of work to get the array semantics fixed in the current codebase---at least much less than any of the full rewrite attempts listed.

**Comment by perimosocordiae** [comment]
If all you want is 2d sparse arrays, then yes, it's easiest to just change a few methods on the existing spmatrix implementation.

Supporting 1d sparse arrays will require a bit more effort, as the assumptions about spmatrix shape are pretty deeply ingrained.

**Comment by zerothi** [comment]
I was also in the need for a sparse matrix with multiple values per sparse element.
Thus I implemented this class:
https://github.com/zerothi/sisl/blob/master/sisl/sparse.py

which allows such matrices. It is entirely written in python but is surprisingly fast.

I know it is not entirely what seems suggested here. But could provide ideas.

Also, it is in CSR format.

**Comment by mrocklin** [comment]
Just to be clear the scope of this issue is limited to just "small modifications of the existing scipy.sparse matrix classes to support the ndarray interface where it makes sense.  If people want to discuss alternative implementations of sparse arrays then they should probably create a new issue and not comment here.

**Comment by zerothi** [comment]
Ok, sorry for the confusion.

**Comment by rgommers** [comment]
The README at https://github.com/mrocklin/sparse now says it supports n-D arrays.

**Comment by mrocklin** [comment]
There is still significant value to a scipy.sparse ndarray interface.

1.  This would support layouts like CSR
2.  Numpy could drop the matrix class (mrocklin/sparse depends on scipy.sparse)

**Comment by rgommers** [comment]
Yes, agreed.

It would probably be good to explain in the README of your package what COO is/isn't good for. It now looks like it's generally useful as *the* sparse ndarray implementation.

**Comment by mrocklin** [comment]
I agree that this sounds like a great idea.  I would only want to remove
the y in "your" :)

**Comment by rgommers** [comment]
removed :)

**Comment by rgommers** [comment]
@hameerabbasi that's not being naive, we'd really like to have sparse arrays in `scipy.sparse`. I don't think COO alone is going to fill all needs, but it's a good start. There's a discussion somewhere about which storage formats make sense (can't find it right now), IIRC there were one or two to drop.

**Comment by hameerabbasi** [comment]
Here's my breakdown:

- CSR and CSC intuitively only make sense for 2-D. The decision to make is whether to convert from COO on the fly or to really keep them. 
- CSR, particularly is important for BLAS operations. The decision to make is whether to extend CSR by (N-1)-d indptr(s) or (N-1)-d indices. I think the latter makes more sense.
- CSC will be the same as CSR, substituting last for first index. 
- COO is general. 
- DOK is general. 
- LIL is general, but it serves the same purpose as DOK, with the only difference being, it's faster if you're not overwriting old elements.
- I would recommend BSR -> BCOO, unless it's BLAS accelerated. 

All of them can be rewritten for N-d, of course.

**Comment by njsmith** [comment]
Possible simplifications to consider:

* Instead of having full-fledged CSR and CSC classes, provide functions to convert 2d COO to and from these formats (represented as a tuple `(indptr, indices, data)`), just for calling into BLAS or similar

* For DOK and LIL, instead of implementing the whole matrix/array interface, you could have a much-smaller builder interface that only allows you to add new values, and then has a `done` method that returns a COO array

* Is DIA really enough better than COO to justify its existence?

* If we need COO and BCOO, can we manage to arrange for these to be the same thing, where COO = BCOO with a 1x1 blocksize?

If we have N array classes, then an M-ary operation has to handle N**M possible combinations of types, so keeping the number down seems like a *big* win.

**Comment by hameerabbasi** [comment]
> Instead of having full-fledged CSR and CSC classes, provide functions to convert 2d COO to and from these formats (represented as a tuple (indptr, indices, data)), just for calling into BLAS or similar

> For DOK and LIL, instead of implementing the whole matrix/array interface, you could have a much-smaller builder interface that only allows you to add new values, and then has a done method that returns a COO array

I wholeheartedly support these two completely. In addition, I think we should limit the operations on DOK to read/write and LIL to write, and all other formats to read-only. We should promote the proper way of doing this and maybe have a "workflow" document explaining how to convert to/from.


> Is DIA really enough better than COO to justify its existence?

DIA doesn't make sense in N-d, only 2-D, unless it's BLAS-accelerated. Maybe for `einsum` of some sort? But otherwise, I don't see a need.

> If we need COO and BCOO, can we manage to arrange for these to be the same thing, where COO = BCOO with a 1x1 blocksize?

Seems doable by simply making the data (N + 1)-d instead of 1-D, where the last N dimensions are the block size (remember, the block can be N-d as well). Can be done with minor modifications to the current `sparse.COO` without performance loss, however, converting to/from CSR/CSC I will need to think out.

> If we have N array classes, then an M-ary operation has to handle N**M possible combinations of types, so keeping the number down seems like a big win.

Theoretically, yes. But we only need conversion functions to be correct, and instead of implementing each operation for every possible format, we convert into and return the correct format. And provide an `asformat` of some sort... Which will only also need 2N implementations, one from and one to COO (or whatever we decide is the canonical format).

**Comment by njsmith** [comment]
> Theoretically, yes. But

Just remember that every minute you spend thinking about how to make complexity manageable is a minute you could instead spend on avoiding creating the complexity in the first place :-)

**Comment by mrocklin** [comment]
> For DOK and LIL, instead of implementing the whole matrix/array interface, you could have a much-smaller builder interface that only allows you to add new values, and then has a done method that returns a COO array

+1 on this, although I would probably focus only on DOK

Eventually I wouldn't mind seeing the `scipy.sparse` module move out to a separate package.  This is a commonly needed and well isolated collection of data structures.  It would be nice to decouple them from the sizable `scipy` package.  (this is an idle comment though, I don't plan to do this work personally).

Semi-related, see https://github.com/conda/conda/issues/6756 for a discussion on package size

**Comment by charris** [comment]
> Eventually I wouldn't mind seeing the scipy.sparse module move out to a separate package. 

Or there could just be a separate package with an API that didn't make use of the matrix class. We need to drop matrices as some point and the sparse package makes that difficult.

**Comment by mrocklin** [comment]
I agree with both goals.  I think that they are separate and both good ideas.

**Comment by perimosocordiae** [comment]
This thread has a bunch of different points:

When the time comes to move away from the matrix-like `scipy.sparse` module, I think it would be prudent to give it the `weave` treatment and spin it off into a separate package.

+1 for having a builder interface that eventually outputs a more efficient representation, though I do think it's useful to have a sparse format that admits interleaved updates and operations somewhat efficiently (there's always a tradeoff).

+0 for putting a lot of effort into BCOO. I'm not sure how much usage of block-sparse matrices we have in the wild. Judging by the state of the BSR code and how few bug reports we get for it, I'm inclined to say that it's not very popular.

-1 for routing all operations through a canonical format. This is what `scipy.sparse` did for a long time (and still does in some cases) with CSR format. It makes writing the code much easier, but it also results in significant costs in both time and memory. Worse, it makes it difficult for users to reason about the code (until they start reading the source and following the code paths themselves). Each format we support should have a "butter zone" of methods / operations that are most efficient in that format, and the documentation should reflect that to let users decide which format is most appropriate for their use case.

**Comment by mrocklin** [comment]
Unless someone has a desire to spend significant time on this (and if so, great) I would suggest not really changing any of the existing 2d matrix classes or how they operate, except to drop the matrix interface.  Everything working today should, by virtue of being finished, probably just continue as is.

> -1 for routing all operations through a canonical format. 

However for dimensions greater than 2 things do get weirder, most formats (like CSC/CSR) stop being as useful, so there are really (in my mind) only two viable formats, DOK and COO.  Additionally, when performing computational operations (like `tensordot`) you often convert COO -> COO2D -> CSR/CSC.  So there is a common format here of `COO`, but it's common only because it's the one layout that really makes the mose sense for n > 2, and because there just aren't that many other options.

**Comment by mrocklin** [comment]
@hameerabbasi you've obviously been very busy at `mrocklin/sparse`.  Do you have any interest in also investigating `scipy/sparse` and possibly implementing some of the changes mentioned in this issue?  I don't mean to put you on the spot here, but if someone *is* likely to take on this work near-term then I think that this conversation will quickly become more pragmatic.

**Comment by mrocklin** [comment]
Sparse arrays need a champion in the ecosystem.  Today you seem like the best candidate :)

**Comment by hameerabbasi** [comment]
> +1 for having a builder interface that eventually outputs a more efficient representation, though I do think it's useful to have a sparse format that admits interleaved updates and operations somewhat efficiently (there's always a tradeoff).

Yes, there is. But I've never found use for interleaving... Sparse matrices are mostly build-once, use many times. Of course, if there is a use-case for interleaving, I'll look into it.

> +0 for putting a lot of effort into BCOO. I'm not sure how much usage of block-sparse matrices we have in the wild. Judging by the state of the BSR code and how few bug reports we get for it, I'm inclined to say that it's not very popular.

BCOO is easy... The only difficult part is CSR/CSC conversion, and if I think about it I'm sure I'll come up with a way.

It comes up fairly often in `kron(sparse, dense)` type operations.

>  Do you have any interest in also investigating scipy/sparse and possibly implementing some of the changes mentioned in this issue? I don't mean to put you on the spot here, but if someone is likely to take on this work near-term then I think that this conversation will quickly become more pragmatic.

I'd be honored... But I'd love to have help, obviously. I'd love to follow N-D related `scipy.sparse` issues and work those out in `mrocklin/sparse`. But, if we're going to keep the code separate, It'd be best to move these discussions into separate issues in `mrocklin/sparse`.

**Comment by rgommers** [comment]
> Eventually I wouldn't mind seeing the `scipy.sparse` module move out to a separate package.  This is a commonly needed and well isolated collection of data structures.  It would be nice to decouple them from the sizable `scipy` package.  (this is an idle comment though, I don't plan to do this work personally).

> Semi-related, see https://github.com/conda/conda/issues/6756 for a discussion on package size

This discussion has been coming up semi-regularly for years (splitting up scipy, not specifically just sparse), but the costs in dev time and backwards compatibility impact do not justify the benefits.

Package size isn't strongly correlated with number of modules or API surface by the way, more with Cython and templated C++ usage, and the size of vendored C/Fortran libs.

**Comment by hameerabbasi** [comment]
> Package size isn't strongly correlated with number of modules or API surface by the way, more with Cython and templated C++ usage, and the size of vendored C/Fortran libs.

Semi-related, we will need something like [sortednp](https://gitlab.sauerburger.com/frank/sortednp) good indexing speed and radix sort for a lot of operations, at the very least.

**Comment by mrocklin** [comment]
> This discussion has been coming up semi-regularly for years (splitting up scipy, not specifically just sparse), but the costs in dev time and backwards compatibility impact do not justify the benefits.

I would personally love to see a scipy.sparse module that was split out.  My hope is that the maintenance burden would not be high, although I'm not sure how quickly moving this module has seen in recent years.  If it has been changing a lot recently then I may be wrong about maintenance costs.

Fortunately though we're not stuck with a big decision to make.  Folks can fork off this module and make a separate package in isolation from the scipy package and we can see how well it does.  If it does well then a conversation can happen later about pointing scipy.sparse to that external package.  If it does badly then we just don't do anything.  I hope that we can experiment cheaply here.

Is such an experiment welcome/desired by the scipy developers?

**Comment by josef-pkt** [comment]
IMO, scipy.sparse is much to important and a fundamental numerical tool to drop it from scipy's API and the corresponding within scipy use, sparse.linalg, ...
I don't know much about outsourcing the underlying code, but if it needs to be vendored with scipy, then I guess there wouldn't be a large advantage to it.

**Comment by rgommers** [comment]
> I would personally love to see a scipy.sparse module that was split out.

Including `sparse.csgraph` and `sparse.linalg`? Those together are somewhat isolated, but still used in `scipy.io.matlab`, in `scipy.linalg`, in `scipy.optimize.linprog` and in `scipy.spatial.(c)KDTree`.

So any split-off package would still not be able to make backwards-incompatible changes, and still have to be either depended on or vendored to provide the `scipy.sparse` API.

> I'm not sure how quickly moving this module has seen in recent years. If it has been changing a lot recently then I may be wrong about maintenance costs.

I'd say it has been one of the higher-activity modules.

> Folks can fork off this module and make a separate package in isolation from the scipy package and we can see how well it does. If it does well then a conversation can happen later about pointing scipy.sparse to that external package. If it does badly then we just don't do anything. I hope that we can experiment cheaply here.

Do well in what respect? I'm not sure a split of development resources is helpful. I'd much rather see a focus on sparse arrays as an independent package, and evolving that quickly.

**Comment by hameerabbasi** [comment]
I think what @mrocklin means is, since `numpy.matrix` + subclasses will be deprecated to be removed (#8195, numpy/numpy#10142), we should deprecate the current `scipy.sparse.spmatrix` subclasses, but keep the new sparse package separate. 

It remains to be seen how internal scipy dependencies will be managed, though. If we are removing the current `scipy.sparse.spmatrix` subclasses, then the new package will need to be merged and the dependencies migrated.

**Comment by rgommers** [comment]
We won't deprecate the current sparse matrices (any time soon at least, until well after we have mature sparse arrays), that's unnecessary and the impact would be very high given their much wider usage than `np.matrix` itself. We'll either put `np.matrix` in a separate package and depend on it, or vendor it within scipy.

**Comment by mrocklin** [comment]
> So any split-off package would still not be able to make backwards-incompatible changes, and still have to be either depended on or vendored to provide the scipy.sparse API.

OK, so this seems like a good argument to do the matrix -> ndarray conversion while the module remains in the scipy codebase.

What do you think is the best way to implement classes that operate exactly like the current scipy.sparse matrix classes, but satisfy the ndarray interface (for n <= 2)?  This was the original question of this issue.  Is this something that the scipy project would want to do within itself or should this happen externally?  

If I were to do this I would probably create a new `csr_array` class alongside the `csr_matrix` class, moving most of the logic into `csr_array`, making modifications as necessary.  Then I would have `csr_matrix` inherit from `csr_array` making the reverse changes to ensure consistency with existing code.  I would probably do this with all of the sparse matrix classes.  If this worked well then I might start moving over code within scipy to depend on the new array classes rather than their matrix equivalents, and then finally deprecate the matrix classes.  The full deprecation cycle here would probably take a long while.  This may be more churn than is desired for the scipy project so "no, this probably isn't going to happen" is a fine answer.

**Comment by mrocklin** [comment]
>> I would personally love to see a scipy.sparse module that was split out.

> Including sparse.csgraph and sparse.linalg? Those together are somewhat isolated, but still used in scipy.io.matlab, in scipy.linalg, in scipy.optimize.linprog and in scipy.spatial.(c)KDTree.

> So any split-off package would still not be able to make backwards-incompatible changes, and still have to be either depended on or vendored to provide the scipy.sparse API.

Yeah, if I were to do this outside I would probably start with a vendored version of the existing scipy.sparse code, including things like linalg (I have to admit an ignorance of csgraph).

This might not be wise though, I don't know.  My assumption is that a fresh start would be expensive.  I could be wrong though.

**Comment by ilayn** [comment]
Is there any plan on NumPy side to provide a sparse nD array object? If there is, then we can wait for that to happen instead of duplicating efforts. 

Speaking of ideal situations, if the common view on the roles of NumPy/SciPy is assumed here, I would expect to find the sparse array object inside NumPy namespace (just like regular arrays) and SciPy implementing the linear algebra on top.

Otherwise as mentioned, array/matrix have to coexist until array gets mature enough and the message spreads around as best practice.

**Comment by hameerabbasi** [comment]
@ilayn Not Numpy (AFAIK), but there is an effort over at [mrocklin/sparse](https://github.com/mrocklin/sparse).

But if we're talking about convention, all the sparse formats are in Scipy at this time, and only the base (dense) class `np.matrix` is in Numpy.

**Comment by pv** [comment]
As I see it,

- The current `csr_matrix` et al. probably need to be retained in `scipy.sparse` for a long time to retain backward compatibility.

- The `scipy.sparse` linalg etc. routines can likely be taught to operate with "new" sparse arrays, regardless of whether they are implemented in `scipy.sparse` or in a separate package (which then likely becomes a dependency for scipy). Partial transitioning of the package from matrix->array semantics is probably possible retaining backward compat, but full switch necessarily involves some breakage.

- Developing in a separate package allows faster release cycle than bi-yearly. I don't see a reason why the new sparse matrix package would specifically need to implement everything in scipy.sparse. It could even implement only the data structures, no linear algebra etc. --- in particular, maintaining two versions of scipy.sparse.linalg only because of array vs. matrix semantics in some places sounds like a waste of resources.

- Whether it's a good idea to base new sparse arrays on scipy.sparse internals, probably depends on how ambitious your plan is. The c++/python mixed language implementation adds python overhead for small operations. It may also be better to start with a design that avoids the combinatorial explosion of c++ templates, and moreover not follow the organically grown code structure and aim for implementing all operations with efficiency in mind, rather than in the somewhat piecemeal way in which scipy.sparse does it. On the other hand, if the goal is only to get array semantics for `ndim <= 2`, writing new `scipy.sparse.csr_array` classes is probably not that much work in the end.

**Comment by hameerabbasi** [comment]
> The current `csr_matrix` et al. probably need to be retained in `scipy.sparse` for a long time to retain backward compatibility.

At least 2-3 release cycles, I'd say.

> Partial transitioning of the package from matrix->array semantics is probably possible retaining backward compat, but full switch necessarily involves some breakage.

It does indeed. I assume you're talking about making `scipy.sparse` operate on both `scipy.sparse.spmatrix` classes and the new classes in [mrocklin/sparse](https://github.com/mrocklin/sparse), returning whichever is the input? If so, +1.

> Developing in a separate package allows faster release cycle than bi-yearly. I don't see a reason why the new sparse matrix package would specifically need to implement everything in scipy.sparse. It could even implement only the data structures, no linear algebra etc. --- in particular, maintaining two versions of scipy.sparse.linalg only because of array vs. matrix semantics in some places sounds like a waste of resources.

+1 here. We already provide `tocsr()` and `tocsc()` methods that Scipy could use (or copy from our code into the Scipy codebase). The data structures `COO` and `DOK`) are mostly complete, only need to be optimized (we'll need Cython for that). We use `COO` as the read-only to-be-operated-on structure and `DOK` as the sparse N-D building structure.

> Whether it's a good idea to base new sparse arrays on scipy.sparse internals, probably depends on how ambitious your plan is. The c++/python mixed language implementation adds python overhead for small operations. It may also be better to start with a design that avoids the combinatorial explosion of c++ templates, and moreover not follow the organically grown code structure and aim for implementing all operations with efficiency in mind, rather than in the somewhat piecemeal way in which scipy.sparse does it.

We're mostly optimal with respect to possible big-O efficiency. The areas I want to see us improve are indexing (we're horribly inefficient there) and sorting indices (quicksort/merge sort vs radix sort), as these things combined take up over 75% the time of most operations/use-cases. This is one area we could use help in, as I'm not terribly experienced in Cython, although I could work up something if I put my mind to it. (xref mrocklin/sparse#60 and mrocklin/sparse#62)

> If there's enough interest, I think being ambitious and not basing new code on scipy.sparse internal architecture is probably a better choice.

We only use Scipy for `dot`/`tensordot` currently. Of course, if what you're proposing is agreed upon, those would be moved to Scipy.

> Implementing both scipy.sparse.csr_array and "new" more efficient sparse array package also does not sound so efficient. It would sound better to just focus efforts on a single good implementation the data structures.

+1 on this.

**Comment by hameerabbasi** [comment]
It might be worth laying out exactly what  requirements (ideally, the absolute minimum) the SciPy community needs in pydata/sparse _before_ it would be willing to depend on pydata/sparse. You can see the docs and what we've already done so far.

As discussed here, it's best to put linear algebra, graph theoretic algorithms, etc in Scipy and the `ndarray` implementations in `sparse`.

I bring this up, because this is close to being a blocker. Of course, I could work on implementing all formats required for Scipy, but I was wondering if it would be better to start parallel work.

I would be willing to prioritize the required features and also to help in implementing/porting anything here to the new interface.

**Comment by mrocklin** [comment]
> It might be worth laying out exactly what requirements (ideally, the absolute minimum) the SciPy community needs in pydata/sparse before it would be willing to depend on pydata/sparse. You can see the docs and what we've already done so far.

No exact requirements exist unfortuately.  Software tends to be adopted on a person-by-person basis early on until enough people use it.

> As discussed here, it's best to put linear algebra, graph theoretic algorithms, etc in Scipy and the ndarray implementations in sparse.

I don't anticipate the scipy library devs building algorithms on top of an external data structure any time in the immediate future.

> My idea is to slowly wean people off the current scipy.sparse before deprecating it and making this the new default, so it's best to start early.

I agree that starting slowly is good.  I suspect that the sklearn developers would welcome CSR/CSC implementations for 1 and 2 dimensions that satisfied the ndarray interface.  Looking through the scikit-learn codebase for scipy.sparse provides an informative list of what they use.  It's not a small list. 

Just to set expectations, a goal like "replace existing use of scipy.sparse with new code" is ambitious and will likely take more than a year.  If the goal is to have a large impact on user communities then I would take one of the following two approaches:

1.  Work from within the scipy library so that changes immediately have an effect on next release.  I think that the thing to do here is to modify their existing code to support the ndarray interface.  This would probably be entirely separate from the pydata/sparse work.
2.  Find new user groups who are not well served by scipy.sparse and work with them.  The dask.array and XArray communities are such groups, due to the need for ndarray interface (hence my original interest in this project).  There are presumably others groups in the world.

**Comment by pv** [comment]
I expect the ndarray interface will not be enough to work with in 
practice, but scipy.sparse will need to be able to deal with details of 
the sparse arrays, e.g. conversions to CSR and CSC formats, and back.

**Comment by pv** [comment]
What could be done in near future with relatively little work is 
recognizing the "new" sparse matrix format as in aslinearoperator() 
etc., which then directly enables their use in iterative equation and 
eigensolvers.

This is likely possible to do in a backward-compatible manner, and maybe 
even without adding a hard dependency.

Direct equation solvers generally require a specific matrix format (CSC 
for superlu), for which format conversion and access to the underlying 
data structure is necessary.

**Comment by mrocklin** [comment]
> What could be done in near future with relatively little work is
recognizing the "new" sparse matrix format as in aslinearoperator()
etc., which then directly enables their use in iterative equation and
eigensolvers.

I wonder if it would be possible to use the existence of numpy protocols to do this more generally.  I would love to see more situations like this where we don't condition based on the type of an object, but rather on established protocols.  This has been very helpful in some of the other numpy-like projects like dask.array and xarray.

> Direct equation solvers generally require a specific matrix format (CSC
for superlu), for which format conversion and access to the underlying
data structure is necessary.

I agree that data layout is essential for efficient use by algorithms, particularly those written at a lower level.  My hope is that if we implement an ndarray CSR/CSC that we will satisfy enough of the `scipy.sparse.csc_matrix` interface to trick the algorithms :)  We've found in NumPy that this sometimes requires the removal of explicit type checks or conversions if they exist.  Maybe this is a better option and what @hameerabbasi had in mind.  Write ndarray implementations that can mimic scipy.sparse implementations well enough that scipy.sparse code just works.  @pv given your experience with scipy.sparse do you anticipate difficulty with this approach?

**Comment by pv** [comment]
Matthew Rocklin kirjoitti 21.02.2018 klo 15:05:
> I wonder if it would be possible to use the existence of numpy protocols 
> to do this more generally. I would love to see more situations like this 
> where we don't condition based on the type of an object, but rather on 
> established protocols. This has been very helpful in some of the other 
> numpy-like projects like dask.array and xarray.

If the objects can be identified (e.g. `getattr(obj, '__array_ufunc__', 
None) is not None`), it's not a problem to make aslinearoperator() 
recognize them. Only `operator.matmul`, the shape and dtype needs to be 
known here. I guess such a PR can be written without a problem, so 
that's a low-hanging fruit.

>     Direct equation solvers generally require a specific matrix format (CSC
>     for superlu), for which format conversion and access to the underlying
>     data structure is necessary.
> 
> I agree that data layout is essential for efficient use by algorithms, 
> particularly those written at a lower level. My hope is that if we 
> implement an ndarray CSR/CSC that we will satisfy enough of the 
> |scipy.sparse.csc_matrix| interface to trick the algorithms :) 

The relevant superlu solver here (and the suitesparse solvers in other 
packages) are written in C, and deal with inputs as outputs as raw CSC 
data, so they need to know how to unwrap and rewrap the raw data.

The same probably applies to the sparse.csgraph routines (which have an 
additional complication, if the handling of structural zeros is different).

The remaining routines are then the sparse hstack, vstack, block, diag, 
etc. --- these are probably more difficult to make work with a different 
sparse matrix format in a backward compatible way.

**Comment by hameerabbasi** [comment]
> Just to set expectations, a goal like "replace existing use of scipy.sparse with new code" is ambitious and will likely take more than a year.

I realize this. It's part of the reason why I'm hoping someone will accept this as my PhD proposal. I hope to be finished with my Masters within the year.

> It's not a small list.

But most of it IS available in BLAS/LAPACK/ARPACK etc and has wrappers in Scipy, so we only need the basic data structures to work for now.

> If the objects can be identified (e.g. `getattr(obj, '__array_ufunc__', None) is not None`)

I would suggest `isinstance(obj, sparse.SparseArray)` in general.

**Comment by mrocklin** [comment]
> I would suggest isinstance(obj, sparse.SparseArray) in general.

This poses two problems:

1.  scipy would need to know about the sparse library, which introduces dependency issues
2.  It would only apply to the new sparse library, and not to other libraries that might show up in the future.  

We *wish* that scipy hadn't used type-based checks in the past.  I think that rather than asking it to expand those type-based checks to these new types, we should instead determine protocols that avoid this problem more generally in the future.

**Comment by hameerabbasi** [comment]
> I think that rather than asking it to expand those type-based checks to these new types, we should instead determine protocols that avoid this problem more generally in the future.

Numpy's upcoming `gufunc` protocol is great for this. It supports e.g. `dot`, `svd`, etc. We should re-use that instead of implementing our own. Of course, the implementations of the `gufunc`s themselves cannot be made generic like we did for `ufunc.__call__`.

**Comment by hameerabbasi** [comment]
However, I'd be perfectly willing to write up a small document that proposes sparse specific things, like how to detect the format of a sparse array, how to check if something is a sparse array in an implementation-agnostic way, and update it as we go.

**Comment by mrocklin** [comment]
I think that the operators to which @pv was referring were array specific, not sparse specific.

I do agree that it would be useful to understand if something is CSC-like or CSR-like for specific algorithms.  One approach here would be to make a generic `CSX` format that has an attribute for determining the storage hierarchy of axes (as you've discussed before) and then asking scipy.sparse to adopt that attribute upstream for their csc and csr classes and the appropriate algorithms.

**Comment by perimosocordiae** [comment]
Seems analogous to the C-order / F-order distinction for dense arrays. If you have a "compressed" sparse structure, C-order ~= CSR and F-order ~= CSC.

**Comment by hameerabbasi** [comment]
@perimosocordiae It can be made more general than that, read dask/dask-ml#123.

TL;DR: `CSD` can have some uncompressed and some compressed dimensions. The compressed dimensions are stored in a linear fashion similar to CSR et. al. The uncompressed dimensions are stored in a format similar to `COO`. `CSR` has rows compressed, but not columns. `CSC` has columns compressed, but not rows. Even `COO` is a special case of this; where no axis is compressed.

**Comment by hameerabbasi** [comment]
I've begun work on [a small document](https://github.com/hameerabbasi/sparse-ndarray-protocols) showing what protocols for sparse arrays all implementations have to follow (not yet complete). Issues/PRs welcome.

**Comment by hameerabbasi** [comment]
Is there a Scipy Enhancement Proposal format modeled on PEP/NEP? Do NEPs already cover this? My document was mostly complete, just needed details added. I was thinking of adding it there.

**Comment by rgommers** [comment]
We're just about to add a Scipy Enhancement Proposal section to the docs (or separate repo), triggered by the `optimize` improvements that @andyfaff is working on. @andyfaff do you have something in a working state already that @hameerabbasi could mirror?

**Comment by andyfaff** [comment]
@stsievert and I are working on something at https://github.com/andyfaff/scipy/pull/6.
It still needs a bit of work, but it'll give a good idea of what's going on.

It sure is a lot of work (done with the hope of the PEP being accepted).

**Comment by andyfaff** [comment]
Anything that makes the implementer think a bit more about design is always a good thing.

**Comment by hameerabbasi** [comment]
My document is in a workable state now. I'd appreciate it if other people here looked it over and suggested improvements. https://github.com/hameerabbasi/sparse-ndarray-protocols

**Comment by mrocklin** [comment]
General thoughts are that I like the document.  I especially like spelling out the expectations of CSR/CSC/COO, etc. [here](https://github.com/hameerabbasi/sparse-ndarray-protocols#proposed-formats).  As with all proposed specs I'm trying to find things that can be cut and still achieve the desired shared functionality.  To do that, I think we need to decide what the desired shared functionality is.  I think that the goal is to enable sparse array implementations to play nicely with the scipy codebase, the broader ecosystem, and each other, without explicit collaboration.  Concretely this depends on the following: 

1.  Follows all of the normal array things, for which I think we defer to upstream numpy protocols when possible (like `__array_ufunc__`)
2.  Can be used by the more advanced scipy.sparse algorithms (linalg, graphs, ...)
3.  Can work with each other for operations like dot, and such

Looking at the existing scipy.sparse code it looks like many of the more advanced algorithms look like the following:

```python
def f(A, b):
    if not isspmatrix_csc(A):
        A = csc_matrix(A)

    access indptr, data, etc..

def G(A, b):
    if not isspmatrix_coo(A):
        A = coo_matrix(A)

    access coords, data
```

So this makes me a little curious if we can just rely on duck types for CSC/CSR/COO/... as you've defined  [here](https://github.com/hameerabbasi/sparse-ndarray-protocols#proposed-formats), as well as `to_csc()`, `to_csr()`, and `to_csc()` methods for downstream classes, and be done.  Does this suffice for interoperability?

**Comment by hameerabbasi** [comment]
> So this makes me a little curious if we can just rely on duck types for CSC/CSR/COO/... as you've defined here, as well as to_csc(), to_csr(), and to_csc() methods for downstream classes, and be done. Does this suffice for interoperability?

My idea is to make downstream implementations independent of SciPy code entirely, and to make SciPy work with and return the right type from a given implementation. Of course, a bit of changes will need to be made to the SciPy codebase, but might as well do a clean-up right?

Here's how this will be done:

```python
if getattr(arr, '__is_sparray__', None)
    if arr.format != 'csc':
        arr = arr.asformat('csc')
    # Other code here
else:
    # Non-sparse handling code here.
```

**Comment by rgommers** [comment]
It would be good to add a rationale for each of the formats. I understand that some formats share a lot of their implementation, but 14 formats (if I counted correctly) is still a very high number.

**Comment by hameerabbasi** [comment]
Yes, strictly speaking there are 14 formats, but classes for most will just be inherited stubs calling their respective super-formats. Things like `CSR` and `CSC` will need minor specialization to work with linear algebraic operations, but that's about all I can think of.

All in all, we just need 4 (!) (I was surprised myself that there were so little, I just got to counting now) formats to cover everything in the document. I've adopted your suggestion and added potential uses for each format, as well as a statement that implementing them is purely optional. BSD and BDOK alone cover 99% of all use-cases.

**Comment by rth** [comment]
Could someone please summarize the current situation on this issue. Is adding sparse ndarray still envisioned as part of scipy roadmap or is all effort now focused on [pydata/sparse](https://github.com/pydata/sparse)?

For instance, for scikit-learn I think the blockers from using pydata/sparse currently would be,
 - lack of native CSR/CSC support: lot of cython code that is written for those formats, and conversion from COO is not enough (cf https://github.com/pydata/sparse/issues/126#issuecomment-384582337 for more details)
 - additional dependence on numba and llvmlite (https://github.com/pydata/sparse/issues/126). Maybe this position would change in the future, not sure.

I couldn't find benchmarks say for dot product comparing `sparse.COO` and `scipy.sparse.coo_matrix`.

The roadmap https://scipy.github.io/devdocs/roadmap.html#implement-sparse-arrays-in-addition-to-sparse-matrices states that at some point the plan is to start depending on pydata/sparse in scipy.sparse. Does that mean that scipy would also depend on numba, as things stand currently?

**Comment by perimosocordiae** [comment]
There are no current plans to add a sparse ndarray class directly to scipy.

**Comment by rgommers** [comment]
- `scipy.sparse.linalg` now supports PyData Sparse arrays as input (since 1.4.0).
- there is indeed no plan to merge PyData Sparse into SciPy, it's good as a standalone project
- PyData Sparse is maturing, but there's still some way to go. Note that it does have a format (CSD) that generalizes CSR/CSC, see https://github.com/pydata/sparse/pull/258.

> The roadmap https://scipy.github.io/devdocs/roadmap.html#implement-sparse-arrays-in-addition-to-sparse-matrices states that at some point the plan is to start depending on pydata/sparse in scipy.sparse. Does that mean that scipy would also depend on numba, as things stand currently?

This is at least a year off I'd say, first PyData Sparse needs to catch up in terms of performance and features, have a stable release, and then we can start considering it. At that point we will probably keep it an optional dependency for quite a while, so installing SciPy does not pull in Numba.
